src/main/kotlin/com/cognifide/gradle/aem/instance/SatisfyTask.kt

```diff
@@ -46,48 +60,53 @@ open class SatisfyTask : SyncTask() {
 
     @TaskAction
     fun satisfy() {
-        val packageGroups = providePackageGroups()
-        satisfyPackagesOnInstances(packageGroups)
-    }
-
-    private fun providePackageGroups(): List<FileGroup> {
-        logger.info("Providing packages from local and remote sources.")
-
-        val fileGroups = packageProvider.filterGroups(groupFilter)
-        val files = fileGroups.flatMap { it.files }
-
-        logger.info("Packages provided (${files.size}).")
-
-        return fileGroups
-    }
-
-    // TODO shouldAwait should be controllable by each PackageGroup.awaitAfter
-    private fun satisfyPackagesOnInstances(packageGroups: List<FileGroup>) {
         for (packageGroup in packageGroups) {
             logger.info("Satisfying group of packages '${packageGroup.name}'.")
 
-            var shouldAwait = false
+            var anyPackageSatisfied = false
 
-            if (config.deployDistributed) {
-                synchronizeInstances({ sync ->
-                    packageGroup.files.onEach {
-                        if (sync.satisfyPackage(it, { sync.distributePackage(it) })) {
-                            shouldAwait = true
+            synchronizeInstances({ sync ->
+                val packageStates = packageGroup.files.fold(mutableMapOf<File, ListResponse.Package?>(), { states, pkg ->
+                    states[pkg] = sync.determineRemotePackage(pkg); states
+                })
+                val anyPackageSatisfiable = packageStates.any {
+                    sync.isSnapshot(it.key) || it.value == null || !it.value!!.installed
+                }
+
+                if (anyPackageSatisfiable) {
+                    packageGroup.initializer(sync)
+                }
+
+                packageStates.forEach { (pkg, state) ->
+                    when {
+                        sync.isSnapshot(pkg) -> {
+                            logger.info("Satisfying package: $pkg (snapshot).")
+                            sync.deployPackage(pkg)
+                            anyPackageSatisfied = true
                         }
-                    }
-                }, Instance.filter(project, config.deployInstanceAuthorName))
-            } else {
-                synchronizeInstances({ sync ->
-                    packageGroup.files.onEach {
-                        if (sync.satisfyPackage(it, { sync.deployPackage(it) })) {
-                            shouldAwait = true
+                        state == null -> {
+                            logger.info("Satisfying package: $pkg (not uploaded).")
+                            sync.deployPackage(pkg)
+                            anyPackageSatisfied = true
+                        }
+                        !state.installed -> {
+                            logger.info("Satisfying package: $pkg (not installed).")
+                            sync.installPackage(state.path)
```

Comment:
@Skejven that's solving issue that you reported, now we are not re-uploading if it was previously uploaded